marshal.go

```diff
@@ -0,0 +1,103 @@
+package putty
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/sha1"
+	"encoding/binary"
+	"fmt"
+	"math/big"
+	"reflect"
+)
+
+func marshal(val interface{}) (data []byte, keySize int, err error) {
+	v := reflect.ValueOf(val).Elem()
+	buf := bytes.NewBuffer([]byte{})
+
+	err = writeField(v, buf)
+	data = buf.Bytes()
+	keySize = buf.Len()
+	return
+}
+
+func addPadding(data []byte) []byte {
+	keySize := len(data)
+	if keySize%aes.BlockSize == 0 {
+		return data
+	}
+	sha := sha1.Sum(data)
+	data = append(data, make([]byte, aes.BlockSize-keySize&(aes.BlockSize-1))...)
+	copy(data[keySize:], sha[:])
+	return data
+}
+
+func writeField(v reflect.Value, dst *bytes.Buffer) error {
+	fieldType := v.Type()
+
+	switch fieldType.Kind() {
+	case reflect.Struct:
+		for i := 0; i < fieldType.NumField(); i++ {
+			if fieldType.Field(i).PkgPath != "" {
+				return fmt.Errorf("struct contains unexported fields")
+			}
+
+			err := writeField(v.Field(i), dst)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	switch fieldType {
+	case reflect.TypeOf(string("")):
+		err := writeString(dst, v.String())
+		if err != nil {
+			return err
+		}
+	case reflect.TypeOf([]byte(nil)):
+		err := writeBytes(dst, v.Bytes())
+		if err != nil {
+			return err
+		}
+	case reflect.TypeOf(new(big.Int)):
+		switch val := (v.Interface()).(type) {
+		case *big.Int:
+			err := writeBigInt(dst, val)
+			if err != nil {
+				return err
+			}
+		default:
+			return fmt.Errorf("unable to set big int")
+		}
+	default:
+		return fmt.Errorf("unknown type %s", fieldType)
+	}
+
+	return nil
+}
+
+func writeBytes(dst *bytes.Buffer, data []byte) error {
+	length := uint32(len(data))
+	// write 4 bytes (data size) of the next element size
+	err := binary.Write(dst, binary.BigEndian, &length)
+	if err != nil {
+		return err
+	}
+
+	_, err = dst.Write(data)
+
+	return err
+}
+
+func writeString(dst *bytes.Buffer, data string) error {
+	return writeBytes(dst, []byte(data))
+}
+
+func writeBigInt(dst *bytes.Buffer, data *big.Int) error {
+	b := data.Bytes()
+	if len(b) > 8 && b[0] >= 128 {
```

Comment:
could you please explain this condition?

Reply:
In the newer putty source it became 

/* Multiprecision integers, in both the SSH-1 and SSH-2 formats. */                                             #define put_mp_ssh1(bs, val) \                              BinarySink_put_mp_ssh1(BinarySink_UPCAST(bs), val)  #define put_mp_ssh2(bs, val) \                              BinarySink_put_mp_ssh2(BinarySink_UPCAST(bs), val)

Reply:
What is happening is big int numbers expressed in big endian byte arrays get a zero start when the first digit has a high first bit...  This is probably meant to mimic a sign but for the big numbers.  I ran into this when comparing golang vs c output

Reply:
Yup.  Found the GoLang comment which dictates that the big package doesn't account for negative numbers.  So, yeah, this padding of a zero is needed to fix a go issue.  The good thing is, if GoLang besides to change to allowing signed big Bytes, the test for a zero will not add an additional zero.  ;)

https://pkg.go.dev/math/big#Int.Bytes

"func (*Int) Bytes ¶
func (x *Int) Bytes() []byte
Bytes returns the absolute value of x as a big-endian byte slice.

To use a fixed length slice, or a preallocated one, use FillBytes."

Reply:
Also:  Thank you for highlighting the function in the conversation to give sectional awareness!  So many times this is not done well.  Kudos!

I put in two additional case statements to bump any uint16 or uint24 up to uint32 because this is what PuTTY seems to use by default.